sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/BeamSqlEnv.java

```diff
@@ -174,8 +106,110 @@ public void executeDdl(String sqlStatement) throws ParseException {
+    }
+
+    public BeamSqlEnvBuilder registerBuiltinUdf(Map<String, List<Method>> methods) {
+      for (Map.Entry<String, List<Method>> entry : methods.entrySet()) {
+        for (Method method : entry.getValue()) {
+          jdbcConnection.getCurrentSchemaPlus().add(entry.getKey(), UdfImpl.create(method));
+        }
+      }
+
+      return this;
+    }
+
+    public BeamSqlEnvBuilder addSchema(String name, TableProvider tableProvider) {
+      jdbcConnection.setSchema(name, tableProvider);
+
+      return this;
+    }
+
+    public BeamSqlEnvBuilder setCurrentSchema(String name) {
+      try {
+        jdbcConnection.setSchema(name);
+      } catch (SQLException e) {
+        throw new RuntimeException(e);
+      }
+
+      return this;
+    }
+
+    /** Register a UDF function which can be used in SQL expression. */
+    public BeamSqlEnvBuilder registerUdf(String functionName, Class<?> clazz, String method) {
+      jdbcConnection.getCurrentSchemaPlus().add(functionName, UdfImpl.create(clazz, method));
+
+      return this;
+    }
+
+    /** Register a UDF function which can be used in SQL expression. */
+    public BeamSqlEnvBuilder registerUdf(String functionName, Class<? extends BeamSqlUdf> clazz) {
+      return registerUdf(functionName, clazz, BeamSqlUdf.UDF_METHOD);
+    }
+
+    public BeamSqlEnvBuilder registerUdf(String functionName, SerializableFunction sfn) {
+      return registerUdf(functionName, sfn.getClass(), "apply");
+    }
+
+    /**
+     * Register a UDAF function which can be used in GROUP-BY expression. See {@link
+     * org.apache.beam.sdk.transforms.Combine.CombineFn} on how to implement a UDAF.
+     */
+    public BeamSqlEnvBuilder registerUdaf(String functionName, Combine.CombineFn combineFn) {
+      jdbcConnection.getCurrentSchemaPlus().add(functionName, new UdafImpl(combineFn));
+
+      return this;
+    }
+
+    /** Load all UDF/UDAF from {@link UdfUdafProvider}. */
+    public BeamSqlEnvBuilder loadUdfUdafFromProvider() {
+      ServiceLoader.<UdfUdafProvider>load(UdfUdafProvider.class)
+          .forEach(
+              ins -> {
+                ins.getBeamSqlUdfs().forEach((udfName, udfClass) -> registerUdf(udfName, udfClass));
+                ins.getSerializableFunctionUdfs()
+                    .forEach((udfName, udfFn) -> registerUdf(udfName, udfFn));
+                ins.getUdafs().forEach((udafName, udafFn) -> registerUdaf(udafName, udafFn));
+              });
+
+      return this;
+    }
+
+    public BeamSqlEnvBuilder loadBeamBuiltinFunctions() {
+      for (BeamBuiltinFunctionProvider provider :
+          ServiceLoader.load(BeamBuiltinFunctionProvider.class)) {
+        registerBuiltinUdf(provider.getBuiltinMethods());
+      }
+
+      return this;
+    }
+
+    public BeamSqlEnvBuilder setPlannerName(String name) {
+      plannerName = name;
+      return this;
+    }
+
+    public BeamSqlEnv build() {
+      if (plannerName.equals("CalcitePlanner")) {
```

Comment:
Like using reflection to load planner?

Reply:
Yes

Reply:
I am not using reflection because different planner needs different initialization. Some other planners might not need FrameworksConfig. Therefore for known planners, it's better to construct them explicitly (otherwise what constructor we expect from reflection).

Allows separate planner be registered by reflection makes sense if there is an immediate use case (someone want register new planner in classpath and they know what constructor they have).

Reply:
I disagree. Reflection is the only way to instantiate something that might not be there during compilation time, as far as I know. Otherwise each and every possible planner has to become a part of Beam source code. E.g. imagine what it would be like if [all](https://github.com/apache/beam/blob/6a596679113de5c0860dd0f8c078b892cc2a849b/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JdbcDriver.java#L148) [kinds](https://github.com/apache/beam/blob/6a596679113de5c0860dd0f8c078b892cc2a849b/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/JdbcFactory.java#L96,L99) of [schema factories](https://github.com/apache/calcite-avatica/blob/4b7eee5bf430b916c7c07897b6f60d2b6b6dabb7/core/src/main/java/org/apache/calcite/avatica/UnregisteredDriver.java#L94) would have to become part of Calcite code. Or pipeline options?

This is our contract how we instantiate the planner. We can require whatever initialization contract we like from the implementations. We can start with empty constructor, framework config, pipeline options, jdbc connection, or anything else and evolve from there.

Reply:
I would prefer to leave it for the future. 

Right now there is only one planner anyway. It's less clear what constructor that new planner should have. I probably can add a list of try: try constructor one, not found, then try constructor two, and so on. But the list of constructor should be put there when we know they are useful (e.g. by seeing more planner).

Also, for any planner that has checked in, they should be initialized directly.

Reply:
But then by the same logic I would prefer to not have this configuration logic at all until we have a concrete use case,. Right now it's not a configuration, it's a hardcoded way of doing things and it's unclear how is it useful.

We can try multiple constructors, but not any random constructors, we should have a documented logic that says which specific constructors we support and in which order they will be called. And we can start with something simple, e.g. default constructor, and then add more ways. It's easy to later add more than to remove something that's being used.

Reply:
@akedin using reflection in `else` to construct instance of planner now.